locator/load_sketch.hh

```diff
@@ -8,87 +8,139 @@
+        bool _has_valid_disk_capacity = true;
+        bool _has_all_tablet_sizes = true;
+
+        node_load(size_t shard_count, uint64_t capacity)
+                : _shards(shard_count)
+                , _du({capacity, 0}) {
+            uint64_t shard_capacity = capacity / shard_count;
             for (shard_id i = 0; i < shard_count; ++i) {
-                _shards[i] = 0;
+                _shards[i].id = i;
+                _shards[i].du.capacity = shard_capacity;
             }
         }
 
-        void update_shard_load(shard_id shard, load_type load_delta) {
-            _load += load_delta;
-
-            auto old_load = _shards[shard];
-            auto new_load = old_load + load_delta;
-            _shards_by_load.erase(shard_load{shard, old_load});
-            _shards[shard] = new_load;
-            _shards_by_load.insert(shard_load{shard, new_load});
+        void update_shard_load(shard_id shard, ssize_t tablet_count_delta, int64_t tablet_size_delta) {
+            _shards_by_load.erase(_shards[shard]);
+            _shards[shard].tablet_count += tablet_count_delta;
+            _shards[shard].du.used += tablet_size_delta;
+            _shards_by_load.insert(_shards[shard]);
+            _du.used += tablet_size_delta;
+            _tablet_count += tablet_count_delta;
         }
 
         void populate_shards_by_load() {
             _shards_by_load.clear();
-            for (shard_id i = 0; i < _shards.size(); ++i) {
-                _shards_by_load.insert(shard_load{i, _shards[i]});
-            }
-        }
-
-        load_type& load() noexcept {
-            return _load;
+            _shards_by_load.insert(_shards.begin(), _shards.end());
         }
 
-        const load_type& load() const noexcept {
-            return _load;
+        // Returns storage utilization for the node
+        load_type get_load() const noexcept {
+            return _du.get_load();
         }
     };
     std::unordered_map<host_id, node_load> _nodes;
     token_metadata_ptr _tm;
+    load_stats_ptr _load_stats;
+    uint64_t _default_tablet_size = service::default_target_tablet_size;
+    uint64_t _minimal_tablet_size = 0;
+
 private:
     tablet_replica_set get_replicas_for_tablet_load(const tablet_info& ti, const tablet_transition_info* trinfo) const {
         // We reflect migrations in the load as if they already happened,
         // optimistically assuming that they will succeed.
         return trinfo ? trinfo->next : ti.replicas;
     }
 
-    future<> populate_table(const tablet_map& tmap, std::optional<host_id> host, std::optional<sstring> only_dc) {
+    std::optional<uint64_t> get_disk_capacity_for_node(host_id node) {
+        if (_load_stats) {
+            if (_load_stats->tablet_stats.contains(node)) {
+                return _load_stats->tablet_stats.at(node).effective_capacity;
+            } else if (_load_stats->capacity.contains(node)) {
+                return _load_stats->capacity.at(node);
```

Comment:
Can _load_stats contain mixed data during rolling upgrade, with some nodes having effective_capacity and some having only the old `capacity`?

If so, I'm worried that load balancer may see false imbalance.